diff --git a/usr/bin/init.sh b/usr/bin/init.sh
--- a/usr/bin/init.sh
+++ b/usr/bin/init.sh
@@ -220,7 +220,8 @@ if [[ "${VPN_ENABLED}" == "yes" ]]; then
 		set +e
 		chown -R "${PUID}":"${PGID}" "/config/wireguard" &> /dev/null
 		exit_code_chown=$?
-		chmod -R 775 "/config/wireguard" &> /dev/null
+		chmod 700 "/config/wireguard" &> /dev/null && \
+			find "/config/wireguard" -maxdepth 1 -type f -exec chmod 600 {} +
 		exit_code_chmod=$?
 		set -e
 
