diff --git a/src/base/bittorrent/sessionimpl.cpp b/src/base/bittorrent/sessionimpl.cpp
index 5f3f101fd..18b351f22 100644
--- a/src/base/bittorrent/sessionimpl.cpp
+++ b/src/base/bittorrent/sessionimpl.cpp
@@ -6595,7 +6595,7 @@ void SessionImpl::updateTrackerEntryStatuses(lt::torrent_handle torrentHandle)
         catch (const std::exception &)
         {
         }
-    });
+    }, -1);
 }
 
 void SessionImpl::handleRemovedTorrent(const TorrentID &torrentID, const QString &partfileRemoveError)
diff --git a/src/base/bittorrent/sessionimpl.h b/src/base/bittorrent/sessionimpl.h
index 0005a70d0..e848ffb84 100644
--- a/src/base/bittorrent/sessionimpl.h
+++ b/src/base/bittorrent/sessionimpl.h
@@ -501,9 +501,9 @@ namespace BitTorrent
         }
 
         template <typename Func>
-        void invokeAsync(Func &&func)
+        void invokeAsync(Func &&func, const int priority = 0)
         {
-            m_asyncWorker->start(std::forward<Func>(func));
+            m_asyncWorker->start(std::forward<Func>(func), priority);
         }
 
         bool isAddTrackersFromURLEnabled() const override;
diff --git a/src/webui/api/authcontroller.cpp b/src/webui/api/authcontroller.cpp
index b393f7d4d..914a5f6a8 100644
--- a/src/webui/api/authcontroller.cpp
+++ b/src/webui/api/authcontroller.cpp
@@ -44,14 +44,14 @@ void AuthController::loginAction()
 {
     if (m_sessionManager->session())
     {
-        setStatus(APIStatus::Ok);
+        setResult(u"Ok."_s);
         return;
     }
 
     if (m_sessionManager->validateCredentials(params()[u"username"_s], params()[u"password"_s]))
     {
         m_sessionManager->sessionStart();
-        setStatus(APIStatus::Ok);
+        setResult(u"Ok."_s);
     }
     else
     {
diff --git a/src/webui/api/torrentscontroller.cpp b/src/webui/api/torrentscontroller.cpp
index 752053b43..1aae66fc5 100644
--- a/src/webui/api/torrentscontroller.cpp
+++ b/src/webui/api/torrentscontroller.cpp
@@ -624,14 +624,7 @@ void TorrentsController::infoAction()
         if (!torrentFilter.match(torrent))
             continue;
 
-        QVariantMap serializedTorrent = serialize(*torrent);
-
-        if (includeFiles && torrent->hasMetadata())
-            serializedTorrent.insert(KEY_PROP_FILES, getFiles(torrent));
-        if (includeTrackers)
-            serializedTorrent.insert(KEY_PROP_TRACKERS, getTrackers(torrent));
-
-        torrentList.append(serializedTorrent);
+        torrentList.append(serialize(*torrent));
     }
 
     if (torrentList.isEmpty())
@@ -691,6 +684,24 @@ void TorrentsController::infoAction()
     if ((limit > 0) || (offset > 0))
         torrentList = torrentList.mid(offset, limit);
 
+    if (includeFiles || includeTrackers)
+    {
+        for (QVariant &torrentData : torrentList)
+        {
+            QVariantMap serializedTorrent = torrentData.toMap();
+            const auto id = BitTorrent::TorrentID::fromString(serializedTorrent[KEY_TORRENT_ID].toString());
+            const BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
+            Q_ASSERT(torrent);
+
+            if (includeFiles && torrent->hasMetadata())
+                serializedTorrent.insert(KEY_PROP_FILES, getFiles(torrent));
+            if (includeTrackers)
+                serializedTorrent.insert(KEY_PROP_TRACKERS, getTrackers(torrent));
+
+            torrentData = serializedTorrent;
+        }
+    }
+
     setResult(QJsonArray::fromVariantList(torrentList));
 }
 
diff --git a/src/webui/webapplication.cpp b/src/webui/webapplication.cpp
index 91d9c80fd..0c578418f 100644
--- a/src/webui/webapplication.cpp
+++ b/src/webui/webapplication.cpp
@@ -732,7 +732,23 @@ void WebApplication::sessionInitialize()
     {
         if (!validateBasicAuth(credentials))
             throw UnauthorizedHTTPError();
-        sessionStart();
+
+        m_currentSession = m_sessions.value(m_basicAuthSessionId);
+        if (m_currentSession && m_currentSession->hasExpired(m_sessionTimeout))
+        {
+            delete m_sessions.take(m_basicAuthSessionId);
+            m_currentSession = nullptr;
+        }
+
+        if (m_currentSession)
+        {
+            m_currentSession->updateTimestamp();
+        }
+        else
+        {
+            m_basicAuthSessionId = generateSid();
+            sessionStartImpl(m_basicAuthSessionId, true);
+        }
     }
 }
 
@@ -836,20 +852,30 @@ void WebApplication::sessionStartImpl(const QString &sessionId, const bool useCo
     m_currentSession = new WebSession(sessionId, app());
     m_sessions[m_currentSession->id()] = m_currentSession;
 
-    m_currentSession->registerAPIController(u"app"_s, new AppController(app(), m_currentSession));
-    m_currentSession->registerAPIController(u"clientdata"_s, new ClientDataController(m_clientDataStorage, app(), m_currentSession));
-    m_currentSession->registerAPIController(u"log"_s, new LogController(app(), m_currentSession));
-    m_currentSession->registerAPIController(u"torrentcreator"_s, new TorrentCreatorController(m_torrentCreationManager, app(), m_currentSession));
-    m_currentSession->registerAPIController(u"rss"_s, new RSSController(app(), m_currentSession));
-    m_currentSession->registerAPIController(u"search"_s, new SearchController(app(), m_currentSession));
-    m_currentSession->registerAPIController(u"torrents"_s, new TorrentsController(app(), m_currentSession));
-    m_currentSession->registerAPIController(u"transfer"_s, new TransferController(app(), m_currentSession));
-
-    const auto *btSession = BitTorrent::Session::instance();
-    auto *syncController = new SyncController(app(), m_currentSession);
-    syncController->updateFreeDiskSpace(btSession->freeDiskSpace());
-    connect(btSession, &BitTorrent::Session::freeDiskSpaceChecked, syncController, &SyncController::updateFreeDiskSpace);
-    m_currentSession->registerAPIController(u"sync"_s, syncController);
+    m_currentSession->registerAPIController(u"app"_s, [app = app(), parent = m_currentSession] { return new AppController(app, parent); });
+    m_currentSession->registerAPIController(u"log"_s, [app = app(), parent = m_currentSession] { return new LogController(app, parent); });
+    m_currentSession->registerAPIController(u"rss"_s, [app = app(), parent = m_currentSession] { return new RSSController(app, parent); });
+    m_currentSession->registerAPIController(u"search"_s, [app = app(), parent = m_currentSession] { return new SearchController(app, parent); });
+    m_currentSession->registerAPIController(u"torrents"_s, [app = app(), parent = m_currentSession] { return new TorrentsController(app, parent); });
+    m_currentSession->registerAPIController(u"transfer"_s, [app = app(), parent = m_currentSession] { return new TransferController(app, parent); });
+    m_currentSession->registerAPIController(u"clientdata"_s
+            , [app = app(), parent = m_currentSession, clientDataStorage = m_clientDataStorage]
+    {
+        return new ClientDataController(clientDataStorage, app, parent);
+    });
+    m_currentSession->registerAPIController(u"torrentcreator"_s
+            , [app = app(), parent = m_currentSession, torrentCreationManager = m_torrentCreationManager]
+    {
+        return new TorrentCreatorController(torrentCreationManager, app, parent);
+    });
+    m_currentSession->registerAPIController(u"sync"_s
+            , [app = app(), parent = m_currentSession, btSession = BitTorrent::Session::instance()]
+    {
+        auto *syncController = new SyncController(app, parent);
+        syncController->updateFreeDiskSpace(btSession->freeDiskSpace());
+        connect(btSession, &BitTorrent::Session::freeDiskSpaceChecked, syncController, &SyncController::updateFreeDiskSpace);
+        return syncController;
+    });
 
     if (useCookie)
         setSessionCookie();
@@ -1130,13 +1156,20 @@ void WebSession::setCookieRefreshTime(const std::chrono::seconds timeout)
     m_cookieRefreshTimer.setRemainingTime(time);
 }
 
-void WebSession::registerAPIController(const QString &scope, APIController *controller)
+void WebSession::registerAPIController(const QString &scope, CreateAPIControllerFunc createAPIController)
 {
-    Q_ASSERT(controller);
-    m_apiControllers[scope] = controller;
+    Q_ASSERT(createAPIController);
+    m_apiControllers.insert(scope, {.create = std::move(createAPIController), .instance = nullptr});
 }
 
 APIController *WebSession::getAPIController(const QString &scope) const
 {
-    return m_apiControllers.value(scope);
+    const auto iter = m_apiControllers.find(scope);
+    if (iter == m_apiControllers.end())
+        return nullptr;
+
+    if (!iter->instance)
+        iter->instance = std::invoke(iter->create);
+
+    return iter->instance;
 }
diff --git a/src/webui/webapplication.h b/src/webui/webapplication.h
index 3dedda7b9..feb64d69a 100644
--- a/src/webui/webapplication.h
+++ b/src/webui/webapplication.h
@@ -30,6 +30,7 @@
 #pragma once
 
 #include <chrono>
+#include <functional>
 #include <type_traits>
 #include <utility>
 
@@ -72,6 +73,8 @@ namespace BitTorrent
 class WebSession final : public ApplicationComponent<QObject>, public ISession
 {
 public:
+    using CreateAPIControllerFunc = std::function<APIController *()>;
+
     explicit WebSession(const QString &sid, IApplication *app);
 
     QString id() const override;
@@ -81,14 +84,20 @@ public:
     bool shouldRefreshCookie() const;
     void setCookieRefreshTime(std::chrono::seconds timeout);
 
-    void registerAPIController(const QString &scope, APIController *controller);
+    void registerAPIController(const QString &scope, CreateAPIControllerFunc createAPIController);
     APIController *getAPIController(const QString &scope) const;
 
 private:
+    struct APIControllerEntry
+    {
+        CreateAPIControllerFunc create;
+        APIController *instance = nullptr;
+    };
+
     const QString m_sid;
     QElapsedTimer m_timestamp;
     QDeadlineTimer m_cookieRefreshTimer;
-    QMap<QString, APIController *> m_apiControllers;
+    mutable QMap<QString, APIControllerEntry> m_apiControllers;
 };
 
 class WebApplication final : public ApplicationComponent<QObject>
@@ -150,6 +159,7 @@ private:
 
     // Persistent data
     QHash<QString, WebSession *> m_sessions;
+    QString m_basicAuthSessionId;
 
     // Current data
     WebSession *m_currentSession = nullptr;
